src/main/kotlin/com/varabyte/kobweb/intellij/colors/KobwebColorProvider.kt

```diff
@@ -0,0 +1,222 @@
+                }
+
+                callee.isKobwebColorFunction("argb(value: Int)", "argb(value: Long)") -> run {
+                    val args = element.valueArguments.run {
+                        evaluateArguments<Int>(1) ?: evaluateArguments<Long, Int>(1) { it.toInt() }
+                    } ?: return@run
+                    return safeRgbColor(args[0] and 0x00_FF_FF_FF)
+                }
+
+                callee.isKobwebColorFunction("hsl(h: Float, s: Float, l: Float)") -> run {
+                    val args = element.valueArguments.evaluateArguments<Float>(3) ?: return@run
+                    return safeHslColor(args[0], args[1], args[2])
+                }
+
+                callee.isKobwebColorFunction("hsla(h: Float, s: Float, l: Float, a: Float)") -> run {
+                    val args = element.valueArguments.evaluateArguments<Float>(4) ?: return@run
+                    return safeHslColor(args[0], args[1], args[2])
+                }
+            }
+        }
+
+        else -> null
+    }
+
+    return if (currentDepth <= MAX_SEARCH_DEPTH) {
+        nextElement?.let { traceColor(it, currentDepth + 1) }
+    } else null
+}
+
+/**
+ * Pretty much emulates what CTRL-clicking on a reference would do
+ */
+private fun KtSimpleNameExpression.locateSource(): PsiElement? {
+    val resolvedMainReference = this.mainReference.resolve() ?: return null
+    val sourceDescriptor =
+        PsiNavigationSupport.getInstance().getDescriptor(resolvedMainReference) as? OpenFileDescriptor ?: return null
+    val sourcePsiFile = sourceDescriptor.file.toPsiFile(this.project) ?: return null
+    val sourceElementIdentifier = PsiUtilBase.getElementAtOffset(sourcePsiFile, sourceDescriptor.offset)
+    val sourceElement = sourceElementIdentifier.parent
+
+    return sourceElement
+}
+
+/**
+ * Uses the kotlin analysis api, as it can parse the constants much smarter.
+ * It can parse decimal, hex and binary automatically for example.
+ */
+private inline fun <reified Evaluated, reified Mapped> Collection<KtValueArgument>.evaluateArguments(
+    argCount: Int,
+    evaluatedValueMapper: (Evaluated) -> Mapped
+): Array<Mapped>? {
+    val constantExpressions = this.mapNotNull { it.getArgumentExpression() as? KtConstantExpression }
+
+    val evaluatedArguments = constantExpressions.mapNotNull {
+        analyze(it.containingKtFile) {
+            it.evaluate(KtConstantEvaluationMode.CONSTANT_LIKE_EXPRESSION_EVALUATION)?.value as? Evaluated
+        }
+    }
+
+    return if (evaluatedArguments.size != this.size || evaluatedArguments.size != argCount) null
+    else evaluatedArguments.map(evaluatedValueMapper).toTypedArray()
+}
+
+private inline fun <reified Evaluated> Collection<KtValueArgument>.evaluateArguments(argCount: Int) =
+    evaluateArguments<Evaluated, Evaluated>(argCount) { it }
+
+private fun KtNamedFunction.isKobwebColorFunction(vararg functionSignatures: String): Boolean {
```

Comment:
Just thinking out loud -- this logic might be a bit too broad. We might add a companion method to the Color class in the future and not realize we're breaking the IDE plugin.

What I would suggest is having another variable elsewhere, a set:

```
private val KobwebColorFactoryMethodNames = setOf("argb", "rgb", "rgba", "hsl", "hsla")
```

Then check that the function name is in that set, ro else return false.

Totally unnecessary for *today* but makes us future proof without needing to add anything crazy.

Reply:
This is actually not a problem at all, since the acceptable color function signatures have to be passed as a parameter anyway to match specific color functions. So anything that isnt hard coded in will not return true. So basically what you recommended is already implemented

Reply:
Ah I misunderstood. You're right!